app/assets/javascripts/oxalis/view/layouting/layout_persistence.js

```diff
@@ -18,56 +24,137 @@ const localStorageKeys = {
 function readStoredLayoutConfigs() {
   const storedLayoutVersion = localStorage.getItem(localStorageKeys.currentLayoutVersion);
   if (!storedLayoutVersion || disableLayoutPersistance) {
-    return {};
-  }
-  if (currentLayoutVersion > JSON.parse(storedLayoutVersion)) {
-    return {};
+    return defaultLayoutSchema;
   }
   const layoutString = localStorage.getItem(localStorageKeys.goldenWkLayouts);
-  if (layoutString) {
-    try {
-      return JSON.parse(layoutString);
-    } catch (ex) {
-      // This should only happen if someone tinkers with localStorage manually
-      console.warn("Layout config could not be deserialized.");
+  if (!layoutString) {
+    return defaultLayoutSchema;
+  }
+  try {
+    const version = JSON.parse(storedLayoutVersion);
+    const layouts = JSON.parse(layoutString);
+    if (currentLayoutVersion > version) {
+      if (version !== 5) {
+        return defaultLayoutSchema;
+      }
+      // migrate to newst schema
+      const withMulipleLayoutsSchema = {
+        OrthoLayoutView: {
+          "Custom Layout": layouts.OrthoLayoutView || defaultLayouts.OrthoLayoutView,
+          lastActive: "Custom Layout",
+        },
+        VolumeTracingView: {
+          "Custom Layout": layouts.VolumeTracingView || defaultLayouts.VolumeTracingView,
+          lastActive: "Custom Layout",
+        },
+        ArbitraryLayout: {
+          "Custom Layout": layouts.ArbitraryLayout || defaultLayouts.ArbitraryLayout,
+          lastActive: "Custom Layout",
+        },
+        OrthoLayout: {
+          "Custom Layout": layouts.OrthoLayout || defaultLayouts.OrthoLayout,
+          lastActive: "Custom Layout",
```

Comment:
This `lastActive` attribute has some disadvantages in my opinion. Two example symptoms are that a layout must not be named "lastActive" and that "lastActive" has to be ignored when listing layout names. In general, the problem is that OrthoLayout (and co) contain a map from layout name to layout config AND the information what is active. These are two different things which can be stored independently.
For example, you could have:

```
...
OrthoLayout: {...},
LastActiveLayouts: {
  OrthoLayout: "Custom Layout",
  ArbitraryLayout: "Layout X",
}
```

as a format. Do you agree?

Reply:
Sure it was obvious to me that my solution is not so good. I like your idea way more than mine and it makes the code even better :smiley: